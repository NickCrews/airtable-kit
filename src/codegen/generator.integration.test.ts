/**
 * Integration tests for code generator
 * Tests the full flow from schema to generated TypeScript code
 */

import { describe, it, expect } from 'vitest';
import { generateCode } from './index.ts';
import taskBase from '../tests/taskBase.ts';
import { de } from 'zod/locales';

describe('Code Generator Integration', () => {
  it('should match snapshot with no options', async () => {
    const code = await generateCode(taskBase);
    expect(code).toMatchSnapshot();
  });

  it('should match snapshot with filetype=js', async () => {
    const code = await generateCode(taskBase, { filetype: 'js' });
    expect(code).toMatchSnapshot();
  });

  it('should match snapshot with filetype=ts', async () => {
    const code = await generateCode(taskBase, { filetype: 'ts' });
    expect(code).toMatchSnapshot();
  });

  describe('Name Escaping', () => {
    const baseWithBadNames = {
      ...taskBase,
      name: 'My Base!',
      tables: [
        {
          ...taskBase.tables[0],
          name: 'await',
          fields: [
            {
              ...taskBase.tables[0].fields[0],
              name: '1st Field',
            },
          ],
        },
      ],
    };
    it('should work by default', async () => {
      const code = await generateCode(baseWithBadNames, { filetype: 'ts' });
      expect(code).toMatchInlineSnapshot(`
      "/**
       * Auto-generated from Airtable schema
       * Do not edit manually
       */

      export default {
        "id": "appTaskBase",
        "name": "myBase",
        "tables": [
          {
            "id": "tblTasks",
            "name": "_await",
            "fields": [
              {
                "id": "fldName",
                "name": "_1StField",
                "type": "singleLineText",
                "description": "The name of the task"
              }
            ]
          }
        ]
      } as const;
      "
    `);
    });
    it('should work with custom escape function', async () => {
      const escapeFn = (name: string) => `_${name.replace(/\\W/g, '')}_`;
      const code = await generateCode(baseWithBadNames, { filetype: 'ts', escapeIdentifiers: escapeFn });
      expect(code).toMatchInlineSnapshot(`
      "/**
       * Auto-generated from Airtable schema
       * Do not edit manually
       */

      export default {
        "id": "appTaskBase",
        "name": "_My Base!_",
        "tables": [
          {
            "id": "tblTasks",
            "name": "_await_",
            "fields": [
              {
                "id": "fldName",
                "name": "_1st Field_",
                "type": "singleLineText",
                "description": "The name of the task"
              }
            ]
          }
        ]
      } as const;
      "
    `);
    });
    it('should work with escapeIdentifiers=false', async () => {
      const code = await generateCode(baseWithBadNames, { filetype: 'ts', escapeIdentifiers: false });
      expect(code).toMatchInlineSnapshot(`
      "/**
       * Auto-generated from Airtable schema
       * Do not edit manually
       */

      export default {
        "id": "appTaskBase",
        "name": "My Base!",
        "tables": [
          {
            "id": "tblTasks",
            "name": "await",
            "fields": [
              {
                "id": "fldName",
                "name": "1st Field",
                "type": "singleLineText",
                "description": "The name of the task"
              }
            ]
          }
        ]
      } as const;
      "
    `);
    });
  });
});
