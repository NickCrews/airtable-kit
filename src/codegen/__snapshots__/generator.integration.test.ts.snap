// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Code Generator Integration > should match snapshot with filetype=js 1`] = `
"/**
 * Auto-generated from Airtable schema
 * Do not edit manually
 */

export default {
  "id": "appTaskBase",
  "name": "taskTracker",
  "tables": [
    {
      "id": "tblTasks",
      "name": "tasks",
      "fields": [
        {
          "id": "fldName",
          "name": "name",
          "type": "singleLineText",
          "description": "The name of the task"
        },
        {
          "id": "fldStatus",
          "name": "status",
          "type": "singleSelect",
          "description": "The current status of the task",
          "options": {
            "choices": [
              {
                "id": "selTodo",
                "name": "Todo",
                "color": "blueLight2"
              },
              {
                "id": "selInProgress",
                "name": "In Progress",
                "color": "yellowLight2"
              },
              {
                "id": "selDone",
                "name": "Done",
                "color": "greenLight2"
              }
            ]
          }
        },
        {
          "id": "fldPriority",
          "name": "priority",
          "type": "number",
          "description": "The priority level of the task. Lower numbers indicate higher priority.",
          "options": {
            "precision": 0
          }
        },
        {
          "id": "fldDueDate",
          "name": "dueDate",
          "type": "date",
          "description": "The due date of the task"
        },
        {
          "id": "fldCompleted",
          "name": "completed",
          "type": "checkbox",
          "description": "Whether the task is completed"
        },
        {
          "id": "fldNotes",
          "name": "notes",
          "type": "multilineText",
          "description": "Additional notes about the task"
        },
        {
          "id": "fldTags",
          "name": "tags",
          "type": "multipleSelects",
          "description": "Tags associated with the task",
          "options": {
            "choices": [
              {
                "id": "selUrgent",
                "name": "Urgent",
                "color": "redLight2"
              },
              {
                "id": "selImportant",
                "name": "Important",
                "color": "orangeLight2"
              }
            ]
          }
        },
        {
          "id": "fldAttachments",
          "name": "attachments",
          "type": "multipleAttachments",
          "description": "Files attached to the task"
        },
        {
          "id": "fldCreatedAt",
          "name": "createdAt",
          "type": "createdTime",
          "description": "The time when the task was created"
        },
        {
          "id": "fldUpdatedAt",
          "name": "updatedAt",
          "type": "lastModifiedTime",
          "description": "The time when the task was last updated"
        },
        {
          "id": "fldAssignedTo",
          "name": "assignedTo",
          "type": "multipleRecordLinks",
          "description": "Users assigned to the task",
          "options": {
            "linkedTableId": "tblUsers"
          }
        }
      ]
    },
    {
      "id": "tblUsers",
      "name": "users",
      "fields": [
        {
          "id": "fldEmail",
          "name": "email",
          "type": "email"
        },
        {
          "id": "fldFullName",
          "name": "fullName",
          "type": "singleLineText"
        },
        {
          "id": "fldPhone",
          "name": "phone",
          "type": "phoneNumber",
          "description": "please include area code"
        }
      ]
    }
  ]
};
"
`;

exports[`Code Generator Integration > should match snapshot with filetype=ts 1`] = `
"/**
 * Auto-generated from Airtable schema
 * Do not edit manually
 */

export default {
  "id": "appTaskBase",
  "name": "taskTracker",
  "tables": [
    {
      "id": "tblTasks",
      "name": "tasks",
      "fields": [
        {
          "id": "fldName",
          "name": "name",
          "type": "singleLineText",
          "description": "The name of the task"
        },
        {
          "id": "fldStatus",
          "name": "status",
          "type": "singleSelect",
          "description": "The current status of the task",
          "options": {
            "choices": [
              {
                "id": "selTodo",
                "name": "Todo",
                "color": "blueLight2"
              },
              {
                "id": "selInProgress",
                "name": "In Progress",
                "color": "yellowLight2"
              },
              {
                "id": "selDone",
                "name": "Done",
                "color": "greenLight2"
              }
            ]
          }
        },
        {
          "id": "fldPriority",
          "name": "priority",
          "type": "number",
          "description": "The priority level of the task. Lower numbers indicate higher priority.",
          "options": {
            "precision": 0
          }
        },
        {
          "id": "fldDueDate",
          "name": "dueDate",
          "type": "date",
          "description": "The due date of the task"
        },
        {
          "id": "fldCompleted",
          "name": "completed",
          "type": "checkbox",
          "description": "Whether the task is completed"
        },
        {
          "id": "fldNotes",
          "name": "notes",
          "type": "multilineText",
          "description": "Additional notes about the task"
        },
        {
          "id": "fldTags",
          "name": "tags",
          "type": "multipleSelects",
          "description": "Tags associated with the task",
          "options": {
            "choices": [
              {
                "id": "selUrgent",
                "name": "Urgent",
                "color": "redLight2"
              },
              {
                "id": "selImportant",
                "name": "Important",
                "color": "orangeLight2"
              }
            ]
          }
        },
        {
          "id": "fldAttachments",
          "name": "attachments",
          "type": "multipleAttachments",
          "description": "Files attached to the task"
        },
        {
          "id": "fldCreatedAt",
          "name": "createdAt",
          "type": "createdTime",
          "description": "The time when the task was created"
        },
        {
          "id": "fldUpdatedAt",
          "name": "updatedAt",
          "type": "lastModifiedTime",
          "description": "The time when the task was last updated"
        },
        {
          "id": "fldAssignedTo",
          "name": "assignedTo",
          "type": "multipleRecordLinks",
          "description": "Users assigned to the task",
          "options": {
            "linkedTableId": "tblUsers"
          }
        }
      ]
    },
    {
      "id": "tblUsers",
      "name": "users",
      "fields": [
        {
          "id": "fldEmail",
          "name": "email",
          "type": "email"
        },
        {
          "id": "fldFullName",
          "name": "fullName",
          "type": "singleLineText"
        },
        {
          "id": "fldPhone",
          "name": "phone",
          "type": "phoneNumber",
          "description": "please include area code"
        }
      ]
    }
  ]
} as const;
"
`;

exports[`Code Generator Integration > should match snapshot with no options 1`] = `
"/**
 * Auto-generated from Airtable schema
 * Do not edit manually
 */

export default {
  "id": "appTaskBase",
  "name": "taskTracker",
  "tables": [
    {
      "id": "tblTasks",
      "name": "tasks",
      "fields": [
        {
          "id": "fldName",
          "name": "name",
          "type": "singleLineText",
          "description": "The name of the task"
        },
        {
          "id": "fldStatus",
          "name": "status",
          "type": "singleSelect",
          "description": "The current status of the task",
          "options": {
            "choices": [
              {
                "id": "selTodo",
                "name": "Todo",
                "color": "blueLight2"
              },
              {
                "id": "selInProgress",
                "name": "In Progress",
                "color": "yellowLight2"
              },
              {
                "id": "selDone",
                "name": "Done",
                "color": "greenLight2"
              }
            ]
          }
        },
        {
          "id": "fldPriority",
          "name": "priority",
          "type": "number",
          "description": "The priority level of the task. Lower numbers indicate higher priority.",
          "options": {
            "precision": 0
          }
        },
        {
          "id": "fldDueDate",
          "name": "dueDate",
          "type": "date",
          "description": "The due date of the task"
        },
        {
          "id": "fldCompleted",
          "name": "completed",
          "type": "checkbox",
          "description": "Whether the task is completed"
        },
        {
          "id": "fldNotes",
          "name": "notes",
          "type": "multilineText",
          "description": "Additional notes about the task"
        },
        {
          "id": "fldTags",
          "name": "tags",
          "type": "multipleSelects",
          "description": "Tags associated with the task",
          "options": {
            "choices": [
              {
                "id": "selUrgent",
                "name": "Urgent",
                "color": "redLight2"
              },
              {
                "id": "selImportant",
                "name": "Important",
                "color": "orangeLight2"
              }
            ]
          }
        },
        {
          "id": "fldAttachments",
          "name": "attachments",
          "type": "multipleAttachments",
          "description": "Files attached to the task"
        },
        {
          "id": "fldCreatedAt",
          "name": "createdAt",
          "type": "createdTime",
          "description": "The time when the task was created"
        },
        {
          "id": "fldUpdatedAt",
          "name": "updatedAt",
          "type": "lastModifiedTime",
          "description": "The time when the task was last updated"
        },
        {
          "id": "fldAssignedTo",
          "name": "assignedTo",
          "type": "multipleRecordLinks",
          "description": "Users assigned to the task",
          "options": {
            "linkedTableId": "tblUsers"
          }
        }
      ]
    },
    {
      "id": "tblUsers",
      "name": "users",
      "fields": [
        {
          "id": "fldEmail",
          "name": "email",
          "type": "email"
        },
        {
          "id": "fldFullName",
          "name": "fullName",
          "type": "singleLineText"
        },
        {
          "id": "fldPhone",
          "name": "phone",
          "type": "phoneNumber",
          "description": "please include area code"
        }
      ]
    }
  ]
} as const;
"
`;
